core/conn/odbc/src/odbc/nsksrvr/SrvrConnect.cpp

```diff
@@ -530,11 +542,39 @@ static void* SessionWatchDog(void* arg)
                 okToGo = false;
             }
         }
-
-
-		while(!record_session_done && okToGo)
-		{
-			REPOS_STATS repos_stats = repos_queue.get_task();
+        vector< vector<string> > query_list;
+        vector<string> session_start;
+        vector<string> statement_new_query;
+        vector<string> session_stat_aggregation;
+
+        session_start.push_back("upsert into Trafodion.\"_REPOS_\".metric_session_table values");
+        statement_new_query.push_back("insert into Trafodion.\"_REPOS_\".metric_query_table values");
+        session_stat_aggregation.push_back("insert into Trafodion.\"_REPOS_\".metric_query_aggr_table values");
+
+        query_list.push_back(session_start);
+        query_list.push_back(statement_new_query);
+        query_list.push_back(session_stat_aggregation);
+
+        int query_limit = statisticsCacheSize;
+        int time_limit = aggrInterval;
+        //0:None 1:update 2:insert/upsert cache limit 3:achieve timeline
+        int execute_flag = REPOS_EXECUTE_NONE;
+        clock_t time_start = clock();
+        clock_t time_end= clock();
+
+        REPOS_STATS repos_stats;
+        while(!record_session_done && okToGo)
+        {
+            time_start = clock();
+            while(repos_queue.isEmpty() && (((time_end = clock()) - time_start) / 1000000 < time_limit));
```

Comment:
I'm not sure I understand this logic. The "while" loop looks like it will spin until the time limit expires or repos_queue is non-empty. This raises two concerns: 1. Spin loops burn CPU needlessly. It would be better to put a sleep call inside the loop. 2. In the logic below, you check to see if the queue is non-empty and then dequeue something. This does not look thread-safe. Another thread might dequeue something after the repos_queue.isEmpty() call and before the repos_queue.get_task() call.

Reply:
Thanks.  I think you are right that spin loops will burn CPU needlessly. As for your second point, every mxosrvr will have only  one watchdog thread, and only this function will call repos_queue.get_task(). Besides, the whole watchdog function  will be locked at the beginning and unlocked at the end so that I think it is thread-safe and doesn't need to add lock there. Thanks again and look forward to your insight. @DaveBirdsall

Reply:
If you set the statisticsCacheSize to 1, would we preserve the current behavior where mxosrvr publishes the repository records when it is time. For example session start is published as soon as session starts and session end is updated as soon as session ends. Like wise we need aggregate stats every 1 minute. Will the existing behavior be retained by setting statisticsCacheSize to 1 ?

Reply:
Thanks. Of course not. The existing behavior is that as long as there is data in the queue， it will be published into repository immediately. statisticsCacheSize just determines the number of data cached in the queue. if it is set to 1, the data will be sent immediately once it goes into the queue, like  'The existing behavior'. @venkat1m